src/components/sdk/MeetingControls/AudioInputVFControl.tsx

```diff
@@ -0,0 +1,140 @@
+  unmutedIconTitle?: string;
+}
+
+const AudioInputVFControl: React.FC<Props> = ({
+  muteLabel = 'Mute',
+  unmuteLabel = 'Unmute',
+  mutedIconTitle,
+  unmutedIconTitle
+}) => {
+  const meetingManager = useMeetingManager();
+  const [voiceFocusOn, setVoiceFocusOn] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+  const [enableVoiceFocus, setEnableVoiceFocus] = useState(false);
+  const dropdownWithVFOptions = useRef<PopOverItemProps[] | null>(null);
+  const { muted, toggleMute } = useToggleLocalMute();
+  const [device, setDevice] = useState<Device | AudioTransformDevice | null>(meetingManager.selectedAudioInputTransformDevice);
+  const { supportsVoiceFocus, addVoiceFocus } = useVoiceFocus();
+  const audioInputConfig: DeviceConfig = {
+    additionalDevices: true,
+  };
+  const { devices, selectedDevice } = useAudioInputs(audioInputConfig);
+
+  const dropdownOptions: PopOverItemProps[] = devices.map((device) => ({
+    children: <span>{device.label}</span>,
+    checked: isOptionActive(selectedDevice, device.deviceId),
+    onClick: (): Promise<void> =>
+      meetingManager.selectAudioInputDevice(device.deviceId),
+  }));
+
+  useEffect(() => {
+    if (device instanceof VoiceFocusTransformDevice) {
+      setEnableVoiceFocus(true);
+    } else {
+      setEnableVoiceFocus(false);
+    }
+  }, [device]);
+
+  useEffect(() => {
+    meetingManager.subscribeToSelectedAudioInputTransformDevice(setDevice);
+
+    return (): void => {
+      meetingManager.unsubscribeFromSelectedAudioInputTransformDevice(setDevice);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (supportsVoiceFocus) {
+      const vfOption: PopOverItemProps = {
+        children: <><PopOverSeparator />{isLoading && <Spinner />} {enableVoiceFocus ? <span>Disable voice focus</span> : <span>Enable voice focus</span>}</>,
+        checked: enableVoiceFocus,
+        disabled: isLoading,
+        onClick: () => {
+          setVoiceFocusOn(current => !current);
+        },
+      };
+      dropdownOptions?.push(vfOption);
+    }
+    dropdownWithVFOptions.current = dropdownOptions;
+  }, [supportsVoiceFocus, enableVoiceFocus, voiceFocusOn, isLoading, dropdownOptions]);
+
+  useEffect(() => {
+    async function onVFCheckboxChange() {
+      let current;
+      setIsLoading(true);
+      if (voiceFocusOn) {
+        if (typeof (device) === 'string') {
+          const currentDevice = audioInputSelectionToDevice(device);
+          current = await addVoiceFocus(currentDevice);
+        }
+      } else {
+        if (device instanceof VoiceFocusTransformDevice) {
+          current = device.getInnerDevice();
+        }
+      }
+      if (current) {
+        setDevice(current)
+        meetingManager.selectAudioInputDevice(current);
+      }
+      setIsLoading(false);
+    }
+
+    onVFCheckboxChange();
+  }, [voiceFocusOn]);
+
+  useEffect(() => {
+    async function onChoseNewDevice() {
```

Comment:
Should we move this complexity to the `VoiceFocusProvider` since it is already doing this in an function there or I think could do the same by adding a method there?

So, can we add two more variables to `VoiceFocusProvider` value as below:
- Selected inner device on which we add voice focus (or for which we create the transform device).
- It already maintains the `VoiceFocusDevice` so add that to the provider's value and then let the components be dependent on it. I think that way we will have single source of audio transform device in the provider itself. This `AudioInputVFControl` component as call providers methods and then get updated based on new inner and VFdevice.

I think this will also simplify the `onVFCheckboxChange` implementation. Let me know if this is achievable or please correct me if I am missing something.

Reply:
Discussed with Jing on this, there are few reasons not to do so:
Why we cannot move up to VoiceFocusProvider or AudioInputProvider?
- We can support multiple meetings with the same VoiceFocusProvider if we do not move up the state, otherwise we would maintain audio transform device specific to a meeting which does not make much sense here since AudioInputVFControl will maintain it.
- We also have an AudioInputProvider, but again if someone who does not want to use AudioInputVFControl but simply AudioInputControl there would be un-necessary work done in AudioInputProvider so we would like to avoid it.

Reply:
Could you please correct me if I have got the right points @richnew10 ?

Reply:
Yes. `VoiceFocusProvider` essentially wraps the ability to create a VF device. `MeetingManager` provides the ability to choose a transform device. _When_ a device is created or chosen or enabled, and _which_ meetings you use it with, varies widely between customer applications.

Reply:
Thanks Richard for the explanation, resolving this conversation.